React/src/App.jsx

```diff
@@ -0,0 +1,118 @@
+import { useState } from 'react';
+import Scheduler from 'devextreme-react/scheduler';
+import { Popup, Position, ToolbarItem } from 'devextreme-react/popup';
+import { defaultData } from './data.js';
+import { rrulestr, RRule } from 'rrule';
+import 'devextreme/dist/css/dx.common.css';
+import 'devextreme/dist/css/dx.light.css';
+import './App.css';
+
+const App = () => {
+    const [popupVisible, setPopupVisible] = useState(false);
+    const [data] = useState(defaultData);
+
+    const hideInfo = () => {
+        setPopupVisible(false);
+    }
+
+    const closeButtonOptions = {
+        text: 'Close',
+        onClick: hideInfo,
+    };
+
+    const formattedDate = (date) => new Date(date.getFullYear(), date.getMonth(), date.getDate());
+
+    const handleAppointmentActions = (e, appointmentData) => {
+        let startTime = appointmentData.startDate.getTime();
+        const endTime = appointmentData.endDate.getTime();
+        const recurringAppointment = defaultData.filter((appointment) => appointment?.recurrenceRule).map((appointment) => {
+            return {
+                ...appointment,
+                startDate: new Date(appointment.startDate),
+                endDate: new Date(appointment.endDate),
+            };
+        });
+
+        recurringAppointment.find((appointment) => {
+            const recurrenceOptions = rrulestr(appointment.recurrenceRule);
+            const rule = new RRule({
+                freq: recurrenceOptions.options.freq,
+                interval: recurrenceOptions.options.interval,
+                byweekday: recurrenceOptions.options.byweekday,
+                dtstart: appointment?.startDate,
+            })
+            const betweenDate = rule.between(e.component.getStartViewDate(), e.component.getEndViewDate())
+            const appointmentDuration = appointment.endDate.getTime() - appointment.startDate.getTime();
+
+            if (betweenDate.length > 0) {
+                betweenDate.find((date) => {
+                    let recurrentStartTime = date.getTime();
+
+                    if(appointment.allDay) {
+                        recurrentStartTime = recurrentStartTime - 86400000
+                        startTime = formattedDate(new Date(startTime)).getTime()
+                    }
+                    const recurrentEndTime = recurrentStartTime + appointmentDuration;
+
+                    if (
+                        startTime === recurrentStartTime
+                        || startTime > recurrentStartTime && startTime < recurrentEndTime
+                        || endTime > recurrentStartTime && endTime <= recurrentEndTime
+                    ) {
+                        e.cancel = true;
+                        setPopupVisible(true);
+                    }
+                })
+            }
+        })
+    }
```

Comment:
Let's change our algorithm of detecting overlap between a new appointment and existing recurrent appointments:

**1) Let's move logic to separate pure function:**
Take a look at typescript function in the Angular component comment.
Let's place it at ```src/utils/isOverlapRecurrentAppointment.js```

**2) Let's simplify the component code:**
And add memorization with ```useCallback``` hooks.
```suggestion
    const handleAppointmentActions = useCallback((event, newAppointment) => {
        const recurrentAppointments = data.filter((appointment) => appointment?.recurrenceRule)
            .map((appointment) => ({
                ...appointment,
                startDate: new Date(appointment.startDate),
                endDate: new Date(appointment.endDate),
            }));

        for (const recurrentAppointment of recurrentAppointments) {
            const isOverlap = isOverlapRecurrentAppointment(event, recurrentAppointment, newAppointment);

            if (isOverlap) {
                event.cancel = true;
                setPopupVisible(true)
            }
        }
    }, [data]);

    const handleAppointmentAdding = useCallback((event) => {
        handleAppointmentActions(event, event.appointmentData);
    }, [handleAppointmentActions]);

    const handleAppointmentUpdating = useCallback((event) => {
        handleAppointmentActions(event, event.newData);
    }, [handleAppointmentActions]);
```